### PR TITLE
[FLINK-2991] Add Folding State and use in WindowOperator

### PR DESCRIPTION
**Pull request by aljoscha**
This enables efficient incremental aggregation of fold window.

This also adds:
- WindowedStream.apply(initVal, foldFunction, windowFunction)
- AllWindowedStream.apply(initVal, foldFunction, windowFunction)

### CODE REVIEW

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		DataOutputViewStreamWrapper out = new DataOutputViewStreamWrapper(baos);
+		try {
+			writeKeyAndNamespace(out);
+			byte[] key = baos.toByteArray();
+			byte[] valueBytes = db.get(key);
+			if (valueBytes == null) {
+				return stateDesc.getDefaultValue();
+			}
+			return valueSerializer.deserialize(new DataInputViewStreamWrapper(new ByteArrayInputStream(valueBytes)));
+		} catch (IOException|RocksDBException e) {
```
Comment (rmetzger) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java]: whoohooo .. first time I see it in Flink ;)

**Comment by rmetzger** [comment]
You are adding new `.apply()` variants to the API which are not documented with the PR.

**Comment by aljoscha** [comment]
I'll add doc and also fix some other stuff in the doc that wasn't updated.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	/** User-specified fold function */
+	private final FoldFunction<T, ACC> foldFunction;
+
+	/**
+	 * Creates a new {@code RocksDBFoldingState}.
+	 *
+	 * @param keySerializer The serializer for the keys.
+	 * @param namespaceSerializer The serializer for the namespace.
+	 * @param stateDesc The state identifier for the state. This contains name
+	 *                           and can create a default state value.
```
Comment (tillrohrmann) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java]: indentation

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+		FoldingStateDescriptor<T, ACC> stateDesc,
+		File dbPath,
+		String backupPath) {
+		super(keySerializer, namespaceSerializer, dbPath, backupPath);
+		this.stateDesc = requireNonNull(stateDesc);
+		this.valueSerializer = stateDesc.getSerializer();
+		this.foldFunction = stateDesc.getFoldFunction();
+	}
+
+	protected RocksDBFoldingState(TypeSerializer<K> keySerializer,
```
Comment (tillrohrmann) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java]: java docs

File: flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+import org.apache.flink.api.common.functions.FoldFunction;
+import org.apache.flink.api.common.functions.RichFunction;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * {@link StateDescriptor} for {@link ReducingState}. This can be used to create partitioned
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java]: for `FoldingState`

File: flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+		super(name, typeClass, initialValue);
+		this.foldFunction = requireNonNull(foldFunction);
+
+		if (foldFunction instanceof RichFunction) {
+			throw new UnsupportedOperationException("FoldFunction of FoldingState can not be a RichFunction.");
+		}
+	}
+
+	/**
+	 * Creates a new {@code ReducingStateDescriptor} with the given name and default value.
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java]: FoldingStateDescriptor

File: flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+		super(name, typeInfo, initialValue);
+		this.foldFunction = requireNonNull(foldFunction);
+
+		if (foldFunction instanceof RichFunction) {
+			throw new UnsupportedOperationException("FoldFunction of FoldingState can not be a RichFunction.");
+		}
+	}
+
+	/**
+	 * Creates a new {@code ValueStateDescriptor} with the given name and default value.
```
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/state/FoldingStateDescriptor.java]: FoldingStateDescriptor

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/GenericFoldingState.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+ * @param <W> Generic type that extends both the underlying {@code ValueState} and {@code KvState}.
+ */
+public class GenericFoldingState<K, N, T, ACC, Backend extends AbstractStateBackend, W extends ValueState<ACC> & KvState<K, N, ValueState<ACC>, ValueStateDescriptor<ACC>, Backend>>
+	implements FoldingState<T, ACC>, KvState<K, N, FoldingState<T, ACC>, FoldingStateDescriptor<T, ACC>, Backend> {
+
+	private final W wrappedState;
+	private final FoldFunction<T, ACC> foldFunction;
+
+	@SuppressWarnings("unchecked")
+	public GenericFoldingState(ValueState<ACC> wrappedState, FoldFunction<T, ACC> foldFunction) {
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/GenericFoldingState.java]: java docs

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java (hunk 2 above)
Comment (aljoscha) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java]: what is the correct way to indent these?

**Comment by aljoscha** [comment]
I addressed the comments.

File: flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java (hunk 1 above)
Comment (StephanEwen) [inline flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBFoldingState.java]: ah, we have used it before in quite a few tests, actually...

**Comment by StephanEwen** [comment]
Doe it always require a "WindowFunction"? Can one also use only a FoldFunction?

**Comment by aljoscha** [comment]
You can also do just a "fold". The window function is for when you need the meta information about the window because the fold function does not get that.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractStateBackend.java [hunk 8]
```diff
@@ -148,6 +150,18 @@ public void dispose() {
...
+	 * Creates and returns a new {@link FoldingState}.
+	 *
+	 * @param namespaceSerializer TypeSerializer for the state namespace.
+	 * @param stateDesc The {@code StateDescriptor} that contains the name of the state.
+	 *
+	 * @param <N> The type of the namespace.
+	 * @param <T> Type of the values folded into the state
+	 * @param <ACC> Type of the value in the state	 *
+	 */
+	abstract protected <N, T, ACC> FoldingState<T, ACC> createFoldingState(TypeSerializer<N> namespaceSerializer, FoldingStateDescriptor<T, ACC> stateDesc) throws Exception;
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractStateBackend.java]: protected abstract

**Pull request closed without merging**